src/System.Buffers.Primitives/System/Buffers/ReadOnlyBufferReader.cs

```diff
@@ -92,13 +92,14 @@ private void MoveNext()
             var previous = _nextPosition;
             while (_sequence.TryGet(ref _nextPosition, out var memory, true))
             {
-                _currentPosition = previous;
                 _currentSpan = memory.Span;
-                _index = 0;
                 if (_currentSpan.Length > 0)
                 {
+                    _index = 0;
```

Comment:
Hmmmm, what if the last segment of the sequence is empty and we don't clear index? I think this code should be outside if.

Reply:
Then we don't also change the current span and so the index points to the end of the span, and in turn Unread is empty. The problem with the current (before this PR code) is that MoveNext might reset the index to zero, but not reset _currentIndex, which will make the last segment available to read again (unless the reader checks the _isEnd bit).

Reply:
After this change, if the last segment is empty we wouldn't move to it, and GetPosition would have to traverse again. Not a big deal but if we are optimizing re-walking we should do it all way.